### PR TITLE
Add ProjectAssemblyLoadContext

### LINKED ISSUE #401: Microsoft.DotNet.ProjectModel.Loader
Create load context/app domains based on project.json environments. This will be useful for scaffolding and entity framework.

### PR DESCRIPTION
**Pull request by anurse**
Fixes #401 

/cc @piotrpMSFT @davidfowl

### CODE REVIEW

File: src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                NativeLibraryExtensions = new[] { ".dylib" };
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                NativeLibraryExtensions = new[] { ".so" };
+            }
+            NativeLibraryExtensions = new string[0];
```
Comment (victorhurdugaci) [inline src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs]: Shouldn't the last statement be under an `else`?

Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs]: sure, why not

Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs]: Oh, right, heh it's not just messy, it's broken :). It used to be a method that returned a value ;)

File: test/LoadContextTest/project.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿{
+    "version": "1.0.0-*",
+    "description": "LoadContextTest Console Application",
+    "authors": [ "anurse" ],
```
Comment (victorhurdugaci) [inline test/LoadContextTest/project.json]: Clean this file

File: src/Microsoft.DotNet.ProjectModel/Constants.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
```
Comment (victorhurdugaci) [inline src/Microsoft.DotNet.ProjectModel/Constants.cs]: License comment? (for the other files too)

File: test/LoadContextTest/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("LoadContextTest")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("LoadContextTest")]
```
Comment (victorhurdugaci) [inline test/LoadContextTest/Properties/AssemblyInfo.cs]: Clean?

File: src/Microsoft.DotNet.ProjectModel/Constants.cs (hunk 3 above)
Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel/Constants.cs]: Move over @troydai, there's a new License Sherrif in town ;)

File: src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            {
+                NativeLibraryExtensions = new[] { ".dylib" };
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                NativeLibraryExtensions = new[] { ".so" };
+            }
+            else
+            {
+                NativeLibraryExtensions = new string[0];
```
Comment (muratg) [inline src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs]: String.Empty() ?

**Comment by anurse** [comment]
@dotnet-bot Test this please

**Comment by anurse** [comment]
@dotnet-bot test this please

**Comment by mmitche** [comment]
@dotnet-bot test Windows_NT Debug Build please

File: src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    /// An <see cref="AssemblyLoadContext"/> designed to allow loading from a Project.
+    /// </summary>
+    public class ProjectLoadContext : AssemblyLoadContext
+    {
+        private readonly IDictionary<AssemblyName, string> _assemblyPaths;
+        private readonly IDictionary<string, string> _nativeLibraries;
+        private readonly IEnumerable<string> _searchPaths;
+
+        private static readonly string[] NativeLibraryExtensions;
+        private static readonly string[] ManagedAssemblyExtensions = new[]
```
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs]: Can we factor this out to a separate class? this seems like reference data we'll want to use in other places.

File: src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            ".dll",
+            ".ni.dll",
+            ".exe",
+            ".ni.exe"
+        };
+
+
+        static ProjectLoadContext()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs]: This too

File: src/Microsoft.DotNet.ProjectModel/Constants.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DotNet.ProjectModel
+{
+    internal static class Constants
+    {
+        public static readonly string DefaultOutputDirectory = "bin";
```
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.ProjectModel/Constants.cs]: I'd put these in OutputDirectories.Default and Configuration.Default, respectively. It's hard to rebuild context from generalized Constants classes.

**Comment by piotrpMSFT** [comment]
Some small comments. I'm merging for the time being to get downstream features unblocked, but do consider making these changes.

File: src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs (hunk 7 above)
Comment (brthor) [inline src/Microsoft.DotNet.ProjectModel.Loader/ProjectLoadContext.cs]: This same logic exists in Microsoft.DotNet.Cli.Utils.Constants

Please consolidate to one or the other.

**Comment by anurse** [comment]
I'm all for centralizing these constants but do we know where best to put it? It's not something I would expect as a public API from Project Model, and there aren't really any other shared places to put it.

**Pull request merged**